master/internal/api_workspace.go

```diff
@@ -312,81 +476,185 @@ func (a *apiServer) PostWorkspace(
 			return nil, status.Error(codes.PermissionDenied, err.Error())
 		}
 	}
+
 	if req.CheckpointStorageConfig != nil && len(req.CheckpointStorageConfig.Fields) > 0 {
 		if err = workspace.AuthZProvider.Get().
 			CanCreateWorkspaceWithCheckpointStorageConfig(ctx, *curUser); err != nil {
 			return nil, status.Error(codes.PermissionDenied, err.Error())
 		}
 	}
 
-	tx, err := db.Bun().BeginTx(ctx, nil)
-	if err != nil {
-		return nil, err
-	}
-	defer func() {
-		if err = tx.Rollback(); err != nil && err != sql.ErrTxDone {
-			log.WithError(err).Error("error rolling back transaction in create workspace")
-		}
-	}()
-
 	w := &model.Workspace{
 		Name: req.Name, UserID: curUser.ID,
 		DefaultComputePool: req.DefaultComputePool, DefaultAuxPool: req.DefaultAuxPool,
 	}
 
-	if req.AgentUserGroup != nil {
-		w.AgentUID = req.AgentUserGroup.AgentUid
-		w.AgentGID = req.AgentUserGroup.AgentGid
-		w.AgentUser = req.AgentUserGroup.AgentUser
-		w.AgentGroup = req.AgentUserGroup.AgentGroup
-	}
+	wsnsBindings := make(map[string]*workspacev1.WorkspaceNamespaceBinding)
 
-	if req.CheckpointStorageConfig != nil && len(req.CheckpointStorageConfig.Fields) > 0 {
-		var bytes []byte
-		bytes, err = req.CheckpointStorageConfig.MarshalJSON()
+	err = db.Bun().RunInTx(ctx, nil, func(ctx context.Context, tx bun.Tx) error {
+		if req.AgentUserGroup != nil {
+			w.AgentUID = req.AgentUserGroup.AgentUid
+			w.AgentGID = req.AgentUserGroup.AgentGid
+			w.AgentUser = req.AgentUserGroup.AgentUser
+			w.AgentGroup = req.AgentUserGroup.AgentGroup
+		}
+
+		if req.CheckpointStorageConfig != nil && len(req.CheckpointStorageConfig.Fields) > 0 {
+			var bytes []byte
+			bytes, err = req.CheckpointStorageConfig.MarshalJSON()
+			if err != nil {
+				return err
+			}
+			var sc expconf.CheckpointStorageConfig
+			w.CheckpointStorageConfig = &sc
+			if err = w.CheckpointStorageConfig.UnmarshalJSON(bytes); err != nil {
+				return err
+			}
+			if err = schemas.IsComplete(w.CheckpointStorageConfig); err != nil {
+				return status.Errorf(codes.InvalidArgument, err.Error())
+			}
+		}
+
+		err = workspace.AddWorkspace(ctx, w, &tx)
 		if err != nil {
-			return nil, err
+			if strings.Contains(err.Error(), db.CodeUniqueViolation) {
```

Comment:
Comparing string contents of error messages is bad practice; is there a way to detect this case through something like `errors.Is`?